### PR TITLE
Keep search and replace texts when changing script

### LINKED ISSUE #27003: Search and replace texts should remain when switching between source files
**Godot version:**
3.1

**OS/device including version:**
Not relevant

**Issue description:**
Search & replace texts should remain when switching between source files.

When you want to verify one by one before replacing, you need to search each file separately.
You search for a method in one node and when trying to search for it in the calling node the search field is empty.
It looks like a bug and very frustrating.

**Steps to reproduce:**

**Minimal reproduction project:**

### PR DESCRIPTION
**Pull request by KoBeWi**
Fixes #27003

For some reason, `popup_replace()` was clearing the replace text, so I had to remove this behavior. Otherwise the replace text was lost when switching to script without replace open.

### CODE REVIEW

**Comment by akien-mga** [comment]
That seems a bit hacky, can't we have just one single `FindReplaceBar`?

**Comment by KoBeWi** [comment]
> can't we have just one single FindReplaceBar?

Well, there's a problem with that. FindReplaceBar is a child of CodeTextEditor, child of ScriptTextEditor, and each tab of script editor has one of them. Not sure where to put the shared FindReplaceBar to make it appear correctly like now.

**Comment by akien-mga** [comment]
IMO the `FindReplaceBar` should have only one instance as child of the `ScriptEditor` (if that's the correct name for the whole "F3" main screen plugin), and another instance in the shader editor plugin. Those two would not need to share their contexts, but the script editor one should be shared by all script tabs.

**Comment by KoBeWi** [comment]
So, like, fixed.

Sadly I couldn't set the shared search_replace_bar as a child of ScriptEditorPlugin due to layout problems. However, I added some methods for moving the search bar between editors. Unfortunately, FindReplaceBar is owned by code_editor, which isn't a property of ScriptEditorBase, so I needed to do some "type code" casts (also changing the location of code_editor to ScriptEditorBase is impossible, because VisualScriptEditor inherits it and doesn't have code_editor).

It's still a bit... hacked, but less than the last time. The only benefit of moving FindReplaceBar to ScriptEditorPlugin would be that it wouldn't be needed to add/remove it between editors. But it still has some references it needs to set, so it wouldn't be much less complex.

**Comment by akien-mga** [comment]
CC @Paulb23

**Comment by akien-mga** [comment]
I'm still quite confused with the various levels of abstraction for the different script editors, but this seems to move a single instance around to different editors, like between the GDScript script editor and the shader editor.

Wouldn't it be possible instead to have one instance for `FindReplaceBar` for each script editor (and thus keep it where it is, without moving it around between editors). I don't think we need to share search and replace texts between unrelated editors, but it should be shared within GDScripts, and similarly shared within shaders.

CC @Paulb23

**Comment by KoBeWi** [comment]
> this seems to move a single instance around to different editors, like between the GDScript script editor and the shader editor.

It's not, they are still separate. ScriptEditorPlugin has a separate code editor for each script, so I'm just passing the reference. It shouldn't matter if the FindReplaceBar is nested inside ScriptEditorPlugin or moved between editors when they need the reference anyways.

File: editor/plugins/script_editor_plugin.cpp [hunk 1]
```diff
@@ -1593,6 +1593,12 @@ void ScriptEditor::ensure_select_current() {
 
 			if (!grab_focus_block && is_visible_in_tree())
 				se->ensure_focus();
+
+			if (Object::cast_to<ScriptTextEditor>(se)) { //sadly this is necessary, because code_editor is not a property of ScriptEditorBase
+				Object::cast_to<ScriptTextEditor>(se)->get_code_editor()->set_find_replace_bar(shared_find_replace_bar);
```
Comment (Paulb23) [inline editor/plugins/script_editor_plugin.cpp]: This should be `Object::cast_to<ScriptTextEditor>(se)->set_find_replace_bar(shared_find_replace_bar);` rather then exposing `CodeEditor`

**Review by Paulb23 — CHANGES_REQUESTED** [review]
The `CodeTextEditor` constructor does the following `find_replace_bar->connect("error", error, "set_text");`. This should be changed to disconnect and connect when passing around the reference. 

I also get the following crash on launch:
```
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] /usr/lib/libc.so.6(+0x3a7e0) [0x7ffff734d7e0] (??:0)
[2] HashMap<StringName, Object::Signal, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::get_element(StringName const&) const (/home/paulb23/development/godot/./core/hash_map.h:186)
[3] HashMap<StringName, Object::Signal, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::getptr(StringName const&) (/home/paulb23/development/godot/./core/hash_map.h:317)
[4] Object::_disconnect(StringName const&, Object*, StringName const&, bool) (/home/paulb23/development/godot/core/object.cpp:1516)
[5] Object::disconnect(StringName const&, Object*, StringName const&) (/home/paulb23/development/godot/core/object.cpp:1512)
[6] FindReplaceBar::set_text_edit(TextEdit*) (/home/paulb23/development/godot/editor/code_editor.cpp:560)
[7] CodeTextEditor::CodeTextEditor() (/home/paulb23/development/godot/editor/code_editor.cpp:1673)
[8] ScriptTextEditor::ScriptTextEditor() (/home/paulb23/development/godot/editor/plugins/script_text_editor.cpp:1711)
[9] /home/paulb23/development/godot/bin/godot.x11.tools.64.llvm() [0x2ca5344] (/home/paulb23/development/godot/editor/plugins/script_text_editor.cpp:1878)
[10] ScriptEditor::edit(Ref<Resource> const&, int, int, bool) (/home/paulb23/development/godot/editor/plugins/script_editor_plugin.cpp:2134)
[11] ScriptEditor::edit(Ref<Resource> const&, bool) (/home/paulb23/development/godot/./editor/plugins/script_editor_plugin.h:427)
[12] ScriptEditor::set_window_layout(Ref<ConfigFile>) (/home/paulb23/development/godot/editor/plugins/script_editor_plugin.cpp:2667)
[13] ScriptEditorPlugin::set_window_layout(Ref<ConfigFile>) (/home/paulb23/development/godot/editor/plugins/script_editor_plugin.cpp:3535)
[14] EditorData::set_plugin_window_layout(Ref<ConfigFile>) (/home/paulb23/development/godot/editor/editor_data.cpp:859)
[15] EditorNode::_load_docks() (/home/paulb23/development/godot/editor/editor_node.cpp:4119)
[16] EditorNode::_sources_changed(bool) (/home/paulb23/development/godot/editor/editor_node.cpp:577)
[17] MethodBind1<EditorNode, bool>::call(Object*, Variant const**, int, Variant::CallError&) (/home/paulb23/development/godot/./core/method_bind.gen.inc:867)
[18] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/paulb23/development/godot/core/object.cpp:921)
[19] Object::emit_signal(StringName const&, Variant const**, int) (/home/paulb23/development/godot/core/object.cpp:1218)
[20] Object::emit_signal(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) (/home/paulb23/development/godot/core/object.cpp:1274)
[21] EditorFileSystem::_notification(int) (/home/paulb23/development/godot/editor/editor_file_system.cpp:1165)
[22] EditorFileSystem::_notificationv(int, bool) (/home/paulb23/development/godot/./editor/editor_file_system.h:108)
[23] Object::notification(int, bool) (/home/paulb23/development/godot/core/object.cpp:931)
[24] SceneTree::_notify_group_pause(StringName const&, int) (/home/paulb23/development/godot/scene/main/scene_tree.cpp:957)
[25] SceneTree::idle(float) (/home/paulb23/development/godot/scene/main/scene_tree.cpp:515)
[26] Main::iteration() (/home/paulb23/development/godot/main/main.cpp:1920)
[27] OS_X11::run() (/home/paulb23/development/godot/platform/x11/os_x11.cpp:3181)
[28] /home/paulb23/development/godot/bin/godot.x11.tools.64.llvm(main+0x1af) [0x1732e3f] (/home/paulb23/development/godot/platform/x11/godot_x11.cpp:56)
[29] /usr/lib/libc.so.6(__libc_start_main+0xf3) [0x7ffff7339ee3] (??:0)
[30] /home/paulb23/development/godot/bin/godot.x11.tools.64.llvm(_start+0x2e) [0x1732bce] (??:?)
-- END OF BACKTRACE --
```

File: editor/code_editor.cpp [hunk 2]
```diff
@@ -527,6 +527,12 @@ String FindReplaceBar::get_replace_text() const {
 	return replace_text->get_text();
 }
 
+void FindReplaceBar::copy_texts_from(FindReplaceBar *p_from) {
```
[the remark this answers is not in the document]
Comment (KoBeWi) [reply editor/code_editor.cpp]: It's a leftover from the first commit. Removed.

**Comment by KoBeWi** [comment]
Added requested changes and resolved some other minor issues. Also
> I also get the following crash on launch:

Doesn't happen to me.

**Comment by aaronfranke** [comment]
@KoBeWi Is this still desired? If so, it needs to be rebased on the latest master branch.

**Comment by KoBeWi** [comment]
Eh, I will just close it for now and rebase/redo if the planned script editor refactor doesn't solve the issue.

**Comment by invisiblesun** [comment]
Why not save the search/replace context (the text data itself) set in a tab in a global context then add code to simply restore that on switch? Either behavior could then be supported as an editor settings.

**Pull request closed without merging**